Proposal: Implementing image vulnerability scanning for tanzu packages

PushkarJ opened the pull request:
Fixes: #2967

PushkarJ committed:
Proposal for implementing image vulnerability scanning for tanzu packages

Signed-off-by: Pushkar Joglekar <pjoglekar@vmware.com>

jpmcb commented on docs/designs/2967-package-image-vulnerability-scanning.md:
```diff
@@ -0,0 +1,635 @@
+* Fail the prow job discretely based on pre-configured rules
+* Alert on failures to identified members of the community
+* Run the vulnerability scan as a periodically running job
+
```
Are there non-goals of this proposal?

PushkarJ replied:
Tried to use Future work as a replacement for non-goals. But perhaps vulnerability disclosure, embargo process falls in neither the goals nor future work. So I can add that as non-goals.

stmcginnis commented on docs/designs/2967-package-image-vulnerability-scanning.md:
Is this set of people expected to ever be anyone beyond VMware employees? If not, it may be better to set up a company ML for this information, further keeping it from any attackers.

PushkarJ replied:
Yes happy to do that and update the wording, since it seems like there is consensus that this will be VMware employees for foreseeable future

nrb commented on docs/designs/2967-package-image-vulnerability-scanning.md:
100% agree that getting as much overlap with upstream as possible is desired here.

I think we can accomplish the desired state without access to prow and testgrid in the short term, but should look to use it in the long term.

PushkarJ replied:
Sounds good. Happy to do it once prow and testgrid is in place unless that is going to take a while and we want to implement this design ASAP.

nrb replied:
I'm really not sure about the prow implementation timeline, so my vote would be to implement in GitHub Actions for the time being. I don't have an opinion on writing this as a script in an Actions runner or using the Trivy action that John mentioned.

PushkarJ replied:
Thanks for the feedback. Talking to @navidshaikh and @rajaskakodkar, who are leading the prow effort, it seems like it would be good to begin with Github Action based workflow for now as prow and testgrid may not be available for several weeks from now. 

I am adding some more info, about how to do this as part of Github Actions. We may not be able to exactly reuse the Github Action for trivy from Aqua Security as is, but with slight modifications can definitely run it as a Github Action to do what we need to do.

jpmcb commented on docs/designs/2967-package-image-vulnerability-scanning.md:
But in theory, all an attacker would need to do is run the commands we use in Prow to get a list of critical CVEs? The barrier to entry for an attacker seems relatively low (which is fine since this is an open source project and anyone anywhere can scan it)

PushkarJ replied:
Yes exactly. They could get the same results, but it would be slightly harder and time consuming for attacker to run this command. So the benefit is small but good enough to not make the vulnerabilities publicly accessible as part of build output.

stmcginnis commented on docs/designs/2967-package-image-vulnerability-scanning.md:
nit: vulnerabilities

stmcginnis commented on docs/designs/2967-package-image-vulnerability-scanning.md:
```diff
@@ -0,0 +1,635 @@
+### Future Work
+
+* Measuring mean time to detect, triage, fix and patch
+* Creating automated Github Issues requesting image bumps from package repos
```
Very trivial nit: s/Github/GitHub/

joshrosso commented:
Overall this would be amazing for the project. Thanks for the well-written proposal.

One question: Would this model (and leveraging prow) have large benefit over [hooking trivy into our harbor registry](https://goharbor.io/docs/2.0.0/administration/vulnerability-scanning/)? I see our introduction of Prow across upstream Tanzu to have a long tail, so wondering if this would be a decent near-term solution?

seemiller commented on docs/designs/2967-package-image-vulnerability-scanning.md:
Would it be more efficient to only run this when a package is added or updated?

jpmcb replied:
> only run this when a package is added or updated?

New CVEs may be discovered and/or published at any time. So one day, a image may be without a CVE and the next, it may have 10 (even tho nothing has changed)

PushkarJ replied:
Yes +1 to what @jpmcb said!

joshrosso commented:
@PushkarJ,, are you involved with the VMware team(s) looking to get prow wired up for all our stuff?

jpmcb commented:
I also see that Trivy has a GitHub action: https://github.com/aquasecurity/trivy-action

Would it be possible to use that with a GitHub `chron` job? This would let us get this started really quickly with little legwork from us.

PushkarJ reviewed (commented):
Thank you all for the initial review!!

PushkarJ commented:
> @PushkarJ,, are you involved with the VMware team(s) looking to get prow wired up for all our stuff?

@joshrosso I missed this reply earlier. I am not involved but happy to join if useful.

PushkarJ commented:
> I also see that Trivy has a GitHub action: https://github.com/aquasecurity/trivy-action
> 
> Would it be possible to use that with a GitHub `chron` job? This would let us get this started really quickly with little legwork from us.

@jpmcb Yes definitely open to this idea. Perhaps it can be a two phased approach. I will also check with VMware team working on prow when they expect it to be available to get a better handle on timelines.

PushkarJ committed:
Review feedback and Github Action updates
- Added alternative / medium term design using Github Actions
- Fixed minor typos
- Added non-goals and updated some wordings based on reviews

- Add trivy scanning in harbor as part of alternatives considered
- Formatting fixes

PushkarJ commented:
@joshrosso , @stmcginnis , @jpmcb @nrb @seemiller Thank you so much, for the review comments so far. Made some updates to the design doc to address the feedback received from you all :) 

- Added Github Actions as a medium term implementation tool to prow and test grid after discussion with VMware RelEng team working on prow (cc @navidshaikh @rajaskakodkar)
- Added reasons on not choosing to hookup trivy with harbor under alternatives considered section
- Fixed minor typos, spellchecks
- Updated wording for more clarity on things like alerting and stories

Hope this addresses all the comments received so far. Happy to work on anything I may have missed.

jpmcb reviewed (commented):
LGTM 👍🏼

joshrosso commented:
@PushkarJ sorry it took us so long to circle back on this, so many things going on in TCE :).

This proposal looks great, we feel ready to get the GitHub action in ASAP and eventually move to the Prow integration, once Tanzu has progressed on its next-gen build automation.

For now, we'll put this in the backlog for our rel-eng team -- @PushkarJ if you have expertise or any extra cycles, we'd also welcome contributions for the initial GH Action. Thanks!
